http_server/main.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+)
+
+type userData struct {
+	username string
+	password string
+	email    string
+}
+
+var (
+
+	// Maps usernames to user's passwords
+	usersMap = make(map[string]userData)
+	// Maps sessions to usernames
+	sessionMap = make(map[string]string)
+)
+
+func welcomeHandler(w http.ResponseWriter, r *http.Request) {
+	names, ok := r.URL.Query()["name"]
+
+	if !ok || len(names[0]) < 1 {
+		http.Error(w, "400 Bad Request", 400)
+		return
+	}
+
+	name := names[0]
+	fmt.Fprintf(w, "<h1>Hello, %s!</h1>", name)
+}
+
+func signUpHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		http.ServeFile(w, r, "signup_form.html")
+	case "POST":
+		signUpPostHandler(w, r)
+	default:
+		fmt.Fprintf(w, "Sent a %s request when GET or POST requests were expected.", r.Method)
+	}
+}
+
+func signUpPostHandler(w http.ResponseWriter, r *http.Request) {
+	if err := r.ParseForm(); err != nil {
+		fmt.Fprintf(w, "Failed to parse the form: %v. ", err)
+		return
+	}
+
+	username := r.FormValue("username")
+	email := r.FormValue("email")
+	password := r.FormValue("password")
+	confirmPassword := r.FormValue("confirmPassword")
+
+	if password != confirmPassword {
+		fmt.Fprint(w, "Submitted passwords differ.")
+		return
+	}
+
+	if _, taken := usersMap[username]; taken {
+		fmt.Fprint(w, "Username already taken, please try again")
+		return
+	}
+
+	usersMap[username] = userData{
+		username: username, password: password, email: email}
+
+	fmt.Fprintf(w, "Successfully signed up, %s! Please sign in", usersMap[username].username)
+}
+
+func signInHandler(w http.ResponseWriter, r *http.Request) {
+
+	switch r.Method {
+	case "GET":
+		fmt.Fprintf(w, `
```

Comment:
Maybe just put it in a separate function for consistency with `signInPostHandler`?